### DISCUSSION TITLE
Çiçekler.

### DISCUSSION THREAD

**Discussion by yukseel** [sonsuzus/sonsuzus#526] [Soru Cevap]
Marmot'un Köstebeği öğle yemeği için yaptığı oyunu gördük... Şimdi Marmot'un akşam yemeği vakti ve hepimizin bildiği gibi Marmot sadece çiçeklerle beslenebiliyor. Akşam yemeklerinde hep kırmızı ve beyaz çiçek yer. Bu yüzden akşam yemeği bazıları beyaz bazıları kırmızı olan birkaç çiçeğin ardışık dizisi olarak gösterilebilir.
Ancak Marmot'un yemeği beğenmesi için bir kural var, Marmot beyaz çiçekleri yalnızca büyüklüğü "k" olan gruplar halinde yemek istiyor.
Şimdi Marmot, a ile b arasında kaç farklı şekilde çiçek yiyebileceğini merak ediyor. Olasılıkların sayısı çok büyük olabilreceğinden sonucu 1000000007 (10^9 + 7) şeklinde yazdırabilirsiniz.

**INPUT---GİRDİ**
İlk satırda t ve k olmak üzere 2 tam sayı bulunur, (1 ≤ t, k ≤ 10^5), burada t kaçıncı test sayısı olduğunu temsil eder.
Sonraki t kadar satır, iki tam sayı olan ai ve bi  (1 ≤ ai ≤ bi ≤ 10^5), bu da i. testi temsil eder.

**OUTPUT---ÇIKTI**
Standart çıktıya t satır yazdırmalısınız. i. satır, Marmot'un akşam yemeğinde ai ile bi arasında kaç farklı şekilde çiçek yiyebileceğini mod 1000000007 (10^9 + 7) ile gösteren sayıyı içermelidir.

**ÖRNEK**
_input:_
`3 2
1 3
2 3
4 4`
_output_
`6
5
5`
**NOTLAR:**
K = 2 ve uzunluk 1 için Marmot (K) yiyebilir. 
K = 2 ve uzunluk 2 için Marmot (KK) ve (BBB) yiyebilir. 
K = 2 ve uzunluk 3 için Marmot (KKK), (KBB) ve (BBK) yiyebilir.
K = 2 ve uzunluk 4 için Marmot, örneğin (BBBB) veya (KBBK) yiyebilir, ama örneğin (BBBK) yiyemez.

**Comment by KaanCanAkyuz** [comment]
```
MOD = 1000000007

def marmot_flower(t, k, queries):
    max_bi = max(b for _, b in queries)
    
    dp = [0] * (max_bi + 1)
    prefix_sum = [0] * (max_bi + 1)
    
    dp[0] = 1  
    for i in range(1, max_bi + 1):
        dp[i] = dp[i - 1]
        if i >= k:
            dp[i] += dp[i - k]
        dp[i] %= MOD
        prefix_sum[i] = (prefix_sum[i - 1] + dp[i]) % MOD
    
    results = []
    for a, b in queries:
        results.append((prefix_sum[b] - prefix_sum[a - 1]) % MOD)
    
    return results

t, k = map(int, input().split())
queries = [tuple(map(int, input().split())) for _ in range(t)]

results = marmot_flower(t, k, queries)
print("\n".join(map(str, results)))
```
